classes/local/pdf/flatten.php

```diff
@@ -0,0 +1,90 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_fileredact\local\pdf;
+
+/**
+ * Flattens the file using ghostscript.
+ *
+ * @package   tool_fileredact
+ * @author    Kevin Pham <kevinpham@catalyst-au.net>
+ * @copyright Catalyst IT, 2022
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class flatten {
+
+    /** @var \stdClass File record */
+    private $filerecord;
+
+    /** @var array additional options (path, contents) provided from the before_file_created hook */
+    private $more;
+
+    /**
+     * Initialise this instance
+     *
+     * @param \stdClass $filerecord
+     * @param array $more
+     */
+    public function __construct(\stdClass $filerecord, array $more) {
+        $this->filerecord = $filerecord;
+        $this->more = $more;
+    }
+
+    /**
+     * Flattens the file using ghostscript
+     *
+     * @return bool as to whether the operation was successful
+     */
+    public function run() {
+        // Get the conversion command.
+        $src = $this->more['pathname'];
+        $temparea = make_request_directory();
+        $dst = $temparea . DIRECTORY_SEPARATOR . $this->filerecord->filename;
+
+        // Prepare the ghostscript (gs) command.
+        $command = $this->get_gs_command($src, $dst);
+
+        // Apply conversion.
+        exec($command);
+
+        // Test to ensure conversion succeeded or not.
+        if (!file_exists($dst)) {
+            // Something has gone wrong in the conversion.
+            return false;
```

Comment:
Id like to see some logging here, probably just the raw shell output of the GS invocation. probably just a debugging() is fine

Reply:
can you just prefix this with 'tool_fileredact:'. Ive run into issues recently where grepping the logs is a pain in the butt unless you know hte exact error message. Instead here, we can grep for tool name :)